File: validate-binary-search-tree/validate-binary-search-tree-mustafasegf.go
package validatebinarysearchtree

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	return driver(root, nil, nil)
}

func driver(root *TreeNode, max, min *int) bool {
	if root == nil {
		return true
	}

	if max != nil && root.Val >= *max {
		return false
	}

	if min != nil && root.Val <= *min {
		return false
	}

	return driver(root.Left, &(root.Val), min) && driver(root.Right, max, &(root.Val))
}

File: validate-binary-search-tree/validate-binary-search-tree-mustafasegf.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn is_valid_bst(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        return Self::rec(&root, i64::MIN, i64::MAX);
    }

    pub fn rec(root: &Option<Rc<RefCell<TreeNode>>>, min: i64, max: i64) -> bool {
        match root {
            Some(node) => {
                let node = node.borrow();
                let val = node.val as i64;

                return val > min
                    && val < max
                    && Self::rec(&node.left, min, val)
                    && Self::rec(&node.right, val, max);
            }
            _ => return true,
        }
    }
}
